### DISCUSSION TITLE
Tableau Cloud cannot connect to MySQL port of ClickHouse using Tableau Bridge on Linux

### DISCUSSION THREAD

**Discussion by burakuyar** [Tab-SE/BridgeCTL#9] [General]
Problem when trying to connect from Tableau Cloud, using Tableau Bridge on RedHat container on Ubuntu host. The structure and connection flow is as below.

- ClickHouse on-prem is up on a Ubuntu instance behind SSH. (That's why Tableau Bridge is needed) ✅ 
- ClickHouse MySQL port is enabled and works as expected when connected from my local via SSH. ✅ 
- Tableau Bridge is up in a Docker container that runs on the instance that hosts ClickHouse. ✅ 
- Tableau Bridge and Tableau Cloud see each other. ✅ 
- When a MySQL connection from Tableau Cloud is initiated, it raises `Can't connect to MySQL server`.

Have you seen such issues before? Is it possible that the container cannot access to the DB on the host of the container?
I've been changing routings but any of them didn't work. Any help or suggestion is much appreciated. Thanks!

**Comment by burakuyar** [comment]
I happened to overcome this issue by modifying the source code a tiny bit before installation (1) and also uploading a custom driver definition file before building the Docker image (2).

1. Needed to convert `network_mode` from `bridge` to `host`. I'd kindly suggest that having the capability to select this on the UI might be a great candidate for the next release, considering that the variables are already in the source code.

2. Needed to upload a Driver Definition File containing the driver below and having the following lines added into the `/etc/odbcinst.ini` (SS from [Tableau Docs](https://www.tableau.com/support/drivers) ).
`https://downloads.mysql.com/archives/get/p/10/file/mysql-connector-odbc-5.3.13-1.el7.x86_64.rpm`
```
[MySQL ODBC 8.0 Unicode Driver]
Driver=/usr/lib64/libmyodbc5w.so
UsageCount=1

[MySQL ODBC 8.0 ANSI Driver]
Driver=/usr/lib64/libmyodbc5a.so
UsageCount=1
```
<img width="1135" alt="image" src="https://github.com/user-attachments/assets/3b873945-13b0-4d58-81e2-7e6b21a94f8c" />


It was a little challenging but wanted to share here to both maybe contribute to the project and also helping to anyone facing similar issues.

Many thanks for the tool!

**Comment by josephflu** [comment]
Thank you for the feedback! We will work on these changes. Hopefully the project will be open source soon so we can take Pull Requests.

**Comment by burakuyar** [comment]
Thanks @josephflu, great news!

As a small addition to the part below, it might be again useful to have the ability to add port mappings while triggering the containers.
> Needed to convert network_mode from bridge to host. I'd kindly suggest that having the capability to select this on the UI might be a great candidate for the next release, considering that the variables are already in the source code.

**Comment by josephflu** [comment]
@burakuyar 

I've added the ability to edit the network_mode from the build bridge page inside the Edit Container DSN mapping dialog. I know this is a bit hidden, but I don't know that it will be used frequently, and it seems slightly related to the DNS mappings fields.  Valid values are 'bridge' or 'host' with 'bridge' as the default. 

<img width="832" alt="image" src="https://github.com/user-attachments/assets/acf95c3a-1ac6-4f34-bbac-e92e896a1be3" />



This has been pushed out as part of version 2.4.18

Let me know if you have any additional suggestions.

**Reply by burakuyar** [reply]
Many thanks for the feature @josephflu . Currently, I'm a little occupied with some other projects, not specifically using Tableau Bridge. However, this feature saves much time for increasing connectivity as needed in different environments. I'll let you know if I observe any other potentially beneficial feature needs.

**Comment by josephflu** [comment]
> I happened to overcome this issue by modifying the source code a tiny bit before installation (1) and also uploading a custom driver definition file before building the Docker image (2).

@burakuyar 
Regarding the driver for clickhouse. I could add the "clickhouse" as a new dropdown in bridgectl build page, but I need to have the exact script for the driver_caddy.yml file. could you write that for me, and I can add it. Let me know if you have questions.

**Reply by burakuyar** [reply]
@josephflu  Regarding this part, I doubt that adding the ClickHouse driver would be useful, as Tableau Cloud does not currently support native ClickHouse drivers.
For now, the only way to consume data from ClickHouse sources is to use MySQL port of ClickHouse servers. And, using MySQL driver in Tableau Bridge was useful for my cases.
